//  File: tao_ca.proto
//  Author: Kevin Walsh <kwalsh@holycross.edu>
//
//  Description: Protocol buffers for TaoCA
//
//  Copyright (c) 2013, Google Inc.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package tao;

import "attestation.proto";  // for Attestation
import "keys.proto";  // for X509Details

enum TaoCARequestType {
  TAO_CA_REQUEST_UNKNOWN = 0;
  TAO_CA_REQUEST_ATTESTATION = 1;  // request Attestation and/or x509 chain
  TAO_CA_REQUEST_SHUTDOWN = 2;  // request server shutdown
}

message TaoCARequest {
  required TaoCARequestType type = 1;

  // An attestation containing a public key. Used when type is
  // TAO_CA_REQUEST_ATTESTATION. A new ROOT attestation will be returned for the
  // same public key, but signed by the policy key.
  optional Attestation attestation = 2;

  // Details for an x509 certificate chain. Optionally used when type is
  // TAO_CA_REQUEST_ATTESTATION. When included in such requests, a new x509
  // certificate chain will be returned for the same public key and the same
  // certificate details, rooted at the policy key.
  optional X509Details x509details = 3;
}

enum TaoCAResponseType {
  TAO_CA_RESPONSE_UNKNOWN = 0;
  TAO_CA_RESPONSE_SUCCESS = 1;
  TAO_CA_RESPONSE_FAILURE = 2;
}

message TaoCAResponse {
  required TaoCAResponseType type = 1;

  // A root attestation for requests of type TAO_CA_REQUEST_CONDENSED_ATTESTATION
  optional Attestation attestation = 2;

  // A PEM-encoded x509 certificate chain
  optional bytes x509chain = 3;

  // An error from the operation
  optional string reason = 4;
}
